tests/mpmc_channel.rs

```diff
@@ -754,6 +754,78 @@ macro_rules! gen_mpmc_tests {
                 assert_eq!(1, elem2.strong_count());
                 assert_eq!(1, elem3.strong_count());
             }
+
+            #[test]
+            fn buffered_uncompleted_recv_does_not_deadlock() {
+                let channel = ChannelType::new();
+                let (waker, count) = new_count_waker();
+                let cx = &mut Context::from_waker(&waker);
+
+                let recv_fut1 = channel.receive();
+                let recv_fut2 = channel.receive();
+                let recv_fut3 = channel.receive();
+                pin_mut!(recv_fut1, recv_fut2, recv_fut3);
+                assert!(recv_fut1.as_mut().poll(cx).is_pending());
+                assert!(recv_fut2.as_mut().poll(cx).is_pending());
+                assert!(recv_fut3.as_mut().poll(cx).is_pending());
+                assert_eq!(count, 0);
+
+                // This wakes the `recv_fut1`.
+                assert_send(cx, &channel, 99);
+                assert_eq!(count, 1);
+
+                assert!(recv_fut2.as_mut().poll(cx).is_pending());
+                assert!(recv_fut3.as_mut().poll(cx).is_pending());
+
+                // This wakes the `recv_fut2`.
+                assert_send(cx, &channel, 420);
+                assert_eq!(count, 2);
+                // This wakes the `recv_fut3`.
+                assert_send(cx, &channel, 69);
+                assert_eq!(count, 3);
+
+                // Once a future is waked up, he is free to pop the value
+                // from the buf at any time.
+                assert_receive_done(cx, &mut recv_fut3, Some(99));
+                assert_receive_done(cx, &mut recv_fut2, Some(420));
+                assert_receive_done(cx, &mut recv_fut1, Some(69));
+            }
+
+            #[test]
+            fn unbuffered_uncompleted_send_does_not_deadlock() {
```

Comment:
The title of this test is a bit misleading. The `send`s  are actually completing. The sender does not call `poll()`  again to get the status, but from the implementation point of view the transaction is completed.

Reply:
Yeah unpolled would be better.